### ISSUE TITLE
Suggesting header without proper define

### ISSUE THREAD

**Issue by svnpenn** [include-what-you-use/include-what-you-use#498]
This file compiles fine as is:

~~~
$ cat q.c
#include <stdio.h>
#include <winsock2.h>
int main() {
  u_char x = 1;
  printf("%d\n", x);
}
~~~

If I run it though include-what-you-use, I get this:

~~~
q.c should add these lines:
#include <sys/types.h>  // for u_char

q.c should remove these lines:
- #include <winsock2.h>  // lines 2-2
~~~

Then if I make those changes the program breaks:

~~~
q.c: In function ‘main’:
q.c:6:3: error: unknown type name ‘u_char’
   u_char x = 1;
   ^~~~~~
~~~

**Comment by kimgr** [comment]
Thanks for the report. This looks like it might be triggered by macro expansion. Are you running iwyu from trunk? If not, which version? Can you up the verbosity level and see if it provides some hints?

**Comment by svnpenn** [comment]
0.8 from here

http://include-what-you-use.org/downloads

**Comment by kimgr** [comment]
OK, that should have the latest macro expansion bugs fixed (there's one more on master, but I don't think it's related to this). Could you post the stderr output from `include-what-you-use -Xiwyu -v7 <your input here>`, so we can see the reasoning behind removing winsock2?

Actually, seeing this now, I think the problem might be a built-in mapping for `u_char` to `sys/types.h`... Yes, that mapping exists. The built-in mappings are very Linux-oriented, so there's probably some naming conflict between Linux system headers and Microsoft's winsock headers.

You could try running with the `--no_default_mappings` switch to get rid of all of them, but that might cause something else to break...

**Comment by svnpenn** [comment]
@kimgr that did it, thanks:

~~~
q.c should add these lines:
#include <_bsd_types.h>  // for u_char

q.c should remove these lines:
- #include <winsock2.h>  // lines 2-2
~~~

**Comment by kimgr** [comment]
@svnpenn -- sorry, it sounds like you're being sarcastic, I'm guessing that didn't help solve the core problem? I'll keep the bug open as a good example of non-portable default mappings.

**Comment by svnpenn** [comment]
@kimgr no, it really did do it - `_bsd_types.h` does have `u_char`, and it is smaller than `winsock2.h` - if I was going to change something I might make a mapping file

~~~
[
  include: [<_bsd_types.h>, private, <winsock2.h>, public]
]
~~~

but otherwise your answer is great

**Comment by kimgr** [comment]
Oh, I see. The perils of written communication :-)

**Comment by jru** [comment]
Labeled as enhancement to remember to improve the default mappings :)
